autoplot version of plot_leaerner_prediction()

jakob-r opened the pull request:
After the discussion on #11 I decided to try out how the autoplot version could look like.
This is another PR to solve #10

Now we can also plot multiple train test splits.

The code is not very clean yet.

jakob-r commented:
### examples 

```r
learner = lrn("classif.rpart", predict_type = "prob")
task2 = tsk("sonar")$select(c("V1", "V2"))
rr2 = resample(task2, learner, rsmp("cv", folds =  3), store_models = TRUE)
autoplot(rr2, type = "prediction")
```
![grafik](https://user-images.githubusercontent.com/1888623/72745509-35232500-3bb0-11ea-96e8-6f8bc353a3d8.png)

```r
task3 = tsk("boston_housing")$select(c("dis"))
rr3 = resample(task3, lrn("regr.ranger", predict_type = "se"), rsmp("bootstrap", repeats = 3), store_models = TRUE)
autoplot(rr3, type = "prediction")
```
![grafik](https://user-images.githubusercontent.com/1888623/72745913-20935c80-3bb1-11ea-8b39-61d9b575e7bd.png)

mllg commented on R/ResampleResult.R:
```diff
@@ -67,6 +67,8 @@ autoplot.ResampleResult = function(object, type = "boxplot", measure = NULL, ...
       autoplot(precrec::evalmod(as_precrec(object)), curvetype = "prc", show_cb = TRUE)
     },
 
+    "prediction" = plot_learner_prediction_resample_result(object$learners[[1]], object, measure, ...),
```
Ugh, the name. Can we just call it `plot_learner_prediction()`, too?

jakob-r replied:
1. The name is only internal.
2. `plot_learner_prediction()` has the arguments `learner` (without model), `task` ... and `plot_learner_prediction_resample_result()` has `learner` (with model), `object` (ResampleResult), ... . We could get rid of the ResampleResult Object if we decide to pass the fold argument and the tain and test ids (if we want to differentiate).

mllg commented on R/ResampleResult.R:
```diff
@@ -80,4 +82,128 @@ fortify.ResampleResult = function(model, data, measure = NULL, ...) {
+    folds_facet = facet_wrap(".id", labeller = as_labeller(facet_labels))
+  } else {
+    folds_facet = NULL
+  }
```
This is basically copy-paste.

jakob-r replied:
There probably is more. Not cleaned up yet.

mllg commented on R/ResampleResult.R:
```diff
@@ -80,4 +82,128 @@ fortify.ResampleResult = function(model, data, measure = NULL, ...) {
+
+  grid_res = predict_grid(learners, task, grid_points = grid_points, expand_range = expand_range)
+
+  if (length(learners) > 1) {
```
Do we need faceting? Would it be okay for the user to just call the plot function multiple times?

jakob-r replied:
I don't have any strong opinions on this. So we would allow to pass an argument e.g. `i` which allows to select the right learner? How would we call this argument? `fold` makes sense for CV but not for bootstrap etc. `resample.iter`, `split`, `split_i`, ...?

mllg commented on R/ResampleResult.R:
```diff
@@ -80,4 +82,128 @@ fortify.ResampleResult = function(model, data, measure = NULL, ...) {
+
+task_data = function(object) {
+  data = lapply(seq_along(object$learners), function(i) {
+    data = object$task$data()
```
What about observation which are neither used in train nor in test?

jakob-r replied:
good point. we do not plot them?

jakob-r commented:
updated plots:
```r
learner = lrn("classif.rpart", predict_type = "prob")
task2 = tsk("sonar")$select(c("V1", "V2"))
rr2 = resample(task2, learner, rsmp("cv", folds =  3), store_models = TRUE)
autoplot(rr2, type = "prediction")
```
![grafik](https://user-images.githubusercontent.com/1888623/72812263-7de6e680-3c61-11ea-999f-933c5bb10530.png)

```r
task3 = tsk("boston_housing")$select(c("age"))
rr3 = resample(task3, lrn("regr.ranger", predict_type = "se", min.node.size = 50), rsmp("bootstrap", repeats = 3), store_models = TRUE)
autoplot(rr3, type = "prediction", predict_sets = c())
```
![grafik](https://user-images.githubusercontent.com/1888623/72812624-21d09200-3c62-11ea-846a-0500fcff3ad2.png)

jakob-r commented:
> It would be nice if we could avoid horizontal scrolling.

IMHO: Inserting line breaks without syntactic meaning should be handled by the viewer.

> I am wondering if there is already too much information in the plot?

The function is just showing what is there. Sometimes it might make sense to reduce the information that goes into the plot. @mllg will wirte a function that allows subsetting the `ResampleResult`.
Furthermore you can reduce the plots to be displayed by defining `predict_sets`.

> Maybe the default could be more conservative?

I switched to the default point size of ggplot2 and the default for `predict_sets` only shows the test set.

> What happens if the user supplies a 100 rep 10 fold CV? Should we by default only plot one rep?

Then the user is crazy. 

> Is the point sizes a little too large?

Switched to default point size.

jakob-r commented:
Note: The small helper function `plot_learner_prediction()` creates a resample object with an holdout with a ratio of 1. This triggers a prediction on an data.frame with 0 rows which not all learners can handle. (See https://github.com/mlr-org/mlr3/issues/421) Options:

1. Remove this helper
2. Change the ratio (back to default)
3. Leave it and merge
4. Wait until mlr3 issue gets fixed and then merge

mllg commented:
> 4. Wait until mlr3 issue gets fixed and then merge

Fixed and merged.

pat-s commented:
> IMHO: Inserting line breaks without syntactic meaning should be handled by the viewer.

Not sure what you mean by "viewer" here. I was referring to the code which is > 100 width so that even in a full sized browser window I had to scroll horizontally to view the code. 

> Then the user is crazy.

My intention was that we detect such cases and "rescue" the user/stop early?

berndbischl commented:
> Not sure what you mean by "viewer" here. I was referring to the code which is > 100 width so that even in a full sized browser window I had to scroll horizontally to view the code.

these discussion need to be resolved in the style guide and not here
